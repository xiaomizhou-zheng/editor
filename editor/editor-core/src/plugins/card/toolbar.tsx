import React from 'react';
import { IntlShape } from 'react-intl-next';
import { EditorState, NodeSelection } from 'prosemirror-state';
import { removeSelectedNode, findDomRefAtPos } from 'prosemirror-utils';
import RemoveIcon from '@atlaskit/icon/glyph/editor/remove';
import UnlinkIcon from '@atlaskit/icon/glyph/editor/unlink';
import CogIcon from '@atlaskit/icon/glyph/editor/settings';
import OpenIcon from '@atlaskit/icon/glyph/shortcut';
import { CardPlatform } from '@atlaskit/smart-card';

import { Command } from '../../types';
import {
  FloatingToolbarHandler,
  FloatingToolbarItem,
} from '../floating-toolbar/types';
import {
  ACTION,
  ACTION_SUBJECT,
  INPUT_METHOD,
  EVENT_TYPE,
  addAnalytics,
  AnalyticsEventPayload,
  ACTION_SUBJECT_ID,
} from '../analytics';
import { linkToolbarMessages, linkMessages } from '../../messages';
import commonMessages from '../../messages';

import { Node } from 'prosemirror-model';
import { hoverDecoration } from '../base/pm-plugins/decoration';
import { changeSelectedCardToText } from './pm-plugins/doc';
import { CardPluginState } from './types';
import { CardOptions } from '@atlaskit/editor-common/card';
import { pluginKey } from './pm-plugins/main';
import { ProviderFactory } from '@atlaskit/editor-common/provider-factory';
import { richMediaClassName } from '@atlaskit/editor-common/styles';
import {
  buildEditLinkToolbar,
  editLink,
  editLinkToolbarConfig,
} from './ui/EditLinkToolbar';
import {
  displayInfoForCard,
  findCardInfo,
  titleUrlPairFromNode,
  appearanceForNodeType,
} from './utils';
import { isSafeUrl } from '@atlaskit/adf-schema';
import { LinkToolbarAppearance } from './ui/LinkToolbarAppearance';
import { messages } from './messages';
import buildLayoutButtons from '../../ui/MediaAndEmbedsToolbar';
import {
  buildOpenedSettingsPayload,
  buildVisitedLinkPayload,
} from '../../utils/linking-utils';
import { LinkPickerOptions } from '../hyperlink/types';
import { FLOATING_TOOLBAR_LINKPICKER_CLASSNAME } from './styles';
import { getFeatureFlags } from '../feature-flags-context';

export const removeCard: Command = (state, dispatch) => {
  if (!(state.selection instanceof NodeSelection)) {
    return false;
  }

  const type = state.selection.node.type.name;
  const payload: AnalyticsEventPayload = {
    action: ACTION.DELETED,
    actionSubject: ACTION_SUBJECT.SMART_LINK,
    actionSubjectId: type as
      | ACTION_SUBJECT_ID.CARD_INLINE
      | ACTION_SUBJECT_ID.CARD_BLOCK,
    attributes: {
      inputMethod: INPUT_METHOD.TOOLBAR,
      displayMode: type as
        | ACTION_SUBJECT_ID.CARD_INLINE
        | ACTION_SUBJECT_ID.CARD_BLOCK,
    },
    eventType: EVENT_TYPE.TRACK,
  };
  if (dispatch) {
    dispatch(addAnalytics(state, removeSelectedNode(state.tr), payload));
  }
  return true;
};

export const visitCardLink: Command = (state, dispatch) => {
  if (!(state.selection instanceof NodeSelection)) {
    return false;
  }

  const { type } = state.selection.node;
  const { url } = titleUrlPairFromNode(state.selection.node);

  // All card links should open in the same tab per https://product-fabric.atlassian.net/browse/MS-1583.
  // We are in edit mode here, open the smart card URL in a new window.
  window.open(url);

  if (dispatch) {
    dispatch(
      addAnalytics(
        state,
        state.tr,
        buildVisitedLinkPayload(
          type.name as
            | ACTION_SUBJECT_ID.CARD_INLINE
            | ACTION_SUBJECT_ID.CARD_BLOCK
            | ACTION_SUBJECT_ID.EMBEDS,
        ),
      ),
    );
  }
  return true;
};

export const openLinkSettings: Command = (state, dispatch) => {
  if (!(state.selection instanceof NodeSelection)) {
    return false;
  }
  window.open('https://id.atlassian.com/manage-profile/link-preferences');
  if (dispatch) {
    const { type } = state.selection.node;
    dispatch(
      addAnalytics(
        state,
        state.tr,
        buildOpenedSettingsPayload(
          type.name as
            | ACTION_SUBJECT_ID.CARD_INLINE
            | ACTION_SUBJECT_ID.CARD_BLOCK
            | ACTION_SUBJECT_ID.EMBEDS,
        ),
      ),
    );
  }
  return true;
};

export const floatingToolbar = (
  cardOptions: CardOptions,
  platform?: CardPlatform,
  linkPickerOptions?: LinkPickerOptions,
): FloatingToolbarHandler => {
  return (state, intl, providerFactory) => {
    const { inlineCard, blockCard, embedCard } = state.schema.nodes;
    const nodeType = [inlineCard, blockCard, embedCard];
    const pluginState: CardPluginState = pluginKey.getState(state);
    if (!(state.selection instanceof NodeSelection)) {
      return;
    }
    const selectedNode = state.selection.node;

    if (!selectedNode) {
      return;
    }

    const isEmbedCard = appearanceForNodeType(selectedNode.type) === 'embed';

    /* add an offset to embeds due to extra padding */
    const toolbarOffset: { offset: [number, number] } | {} = isEmbedCard
      ? {
          offset: [0, 24],
        }
      : {};

    // Applies padding override for when link picker is currently displayed
    const className = pluginState.showLinkingToolbar
      ? FLOATING_TOOLBAR_LINKPICKER_CLASSNAME
      : undefined;

    return {
      title: intl.formatMessage(messages.card),
      className,
      nodeType,
      ...toolbarOffset,
      getDomRef: (view) => {
        const element = findDomRefAtPos(
          view.state.selection.from,
          view.domAtPos.bind(view),
        ) as HTMLElement;
        if (!element) {
          return undefined;
        }
        if (isEmbedCard) {
          return element.querySelector(`.${richMediaClassName}`) as HTMLElement;
        }
        return element;
      },

      items: generateToolbarItems(
        state,
        intl,
        providerFactory,
        cardOptions,
        platform,
        linkPickerOptions,
      ),
      ...(pluginState.showLinkingToolbar ? editLinkToolbarConfig : {}),
      scrollable: pluginState.showLinkingToolbar ? false : true,
    };
  };
};

const unlinkCard = (node: Node, state: EditorState): Command => {
  const displayInfo = displayInfoForCard(node, findCardInfo(state));
  const text = displayInfo.title || displayInfo.url;
  if (text) {
    return changeSelectedCardToText(text);
  }

  return () => false;
};

const buildAlignmentOptions = (
  state: EditorState,
  intl: IntlShape,
): FloatingToolbarItem<Command>[] => {
  return buildLayoutButtons(
    state,
    intl,
    state.schema.nodes.embedCard,
    true,
    true,
  );
};

const generateToolbarItems = (
  state: EditorState,
  intl: IntlShape,
  providerFactory: ProviderFactory,
  cardOptions: CardOptions,
  platform?: CardPlatform,
  linkPicker?: LinkPickerOptions,
) => (node: Node): Array<FloatingToolbarItem<Command>> => {
  const { url } = titleUrlPairFromNode(node);
  let metadata = {};
  if (url && !isSafeUrl(url)) {
    return [];
  } else {
    const { title } = displayInfoForCard(node, findCardInfo(state));
    metadata = {
      url: url,
      title: title,
    };
  }

  const pluginState: CardPluginState = pluginKey.getState(state);

  const currentAppearance = appearanceForNodeType(node.type);

  /* mobile builds toolbar natively using toolbarItems */
  if (pluginState.showLinkingToolbar && platform !== 'mobile') {
    return [
      buildEditLinkToolbar({
        providerFactory,
        linkPicker,
        node,
      }),
    ];
  } else {
    const { inlineCard } = state.schema.nodes;
    const toolbarItems: Array<FloatingToolbarItem<Command>> = [
      {
        id: 'editor.link.edit',
        type: 'button',
        selected: false,
        metadata: metadata,
        title: intl.formatMessage(linkToolbarMessages.editLink),
        showTitle: true,
        testId: 'link-toolbar-edit-link-button',
        onClick: editLink,
      },
      { type: 'separator' },
      {
        id: 'editor.link.openLink',
        type: 'button',
        icon: OpenIcon,
        metadata: metadata,
        className: 'hyperlink-open-link',
        title: intl.formatMessage(linkMessages.openLink),
        onClick: visitCardLink,
      },
      { type: 'separator' },
      ...getUnlinkButtonGroup(state, intl, node, inlineCard),
      {
        type: 'copy-button',
        items: [
          {
            state,
            formatMessage: intl.formatMessage,
            nodeType: node.type,
          },
          { type: 'separator' },
        ],
      },
      ...getSettingsButtonGroup(state, intl),
      {
        id: 'editor.link.delete',
        type: 'button',
        appearance: 'danger',
        icon: RemoveIcon,
        onMouseEnter: hoverDecoration(node.type, true),
        onMouseLeave: hoverDecoration(node.type, false),
        onFocus: hoverDecoration(node.type, true),
        onBlur: hoverDecoration(node.type, false),
        title: intl.formatMessage(commonMessages.remove),
        onClick: removeCard,
      },
    ];

    if (currentAppearance === 'embed') {
      const alignmentOptions = buildAlignmentOptions(state, intl);
      if (alignmentOptions.length) {
        alignmentOptions.push({
          type: 'separator',
        });
      }
      toolbarItems.unshift(...alignmentOptions);
    }
    const { allowBlockCards, allowEmbeds } = cardOptions;

    if ((allowBlockCards || allowEmbeds) && currentAppearance) {
      toolbarItems.unshift(
        {
          type: 'custom',
          fallback: [],
          render: (editorView) => (
            <LinkToolbarAppearance
              key="link-appearance"
              url={url}
              intl={intl}
              currentAppearance={currentAppearance}
              editorView={editorView}
              editorState={state}
              allowEmbeds={allowEmbeds}
              platform={platform}
            />
          ),
        },
        {
          type: 'separator',
        },
      );
    }

    return toolbarItems;
  }
};

const getUnlinkButtonGroup = (
  state: EditorState<any>,
  intl: IntlShape,
  node: Node<any>,
  inlineCard: any,
) => {
  return node.type === inlineCard
    ? ([
        {
          id: 'editor.link.unlink',
          type: 'button',
          title: intl.formatMessage(linkToolbarMessages.unlink),
          icon: UnlinkIcon,
          onClick: unlinkCard(node, state),
        },
        { type: 'separator' },
      ] as Array<FloatingToolbarItem<Command>>)
    : [];
};

const getSettingsButtonGroup = (
  state: EditorState<any>,
  intl: IntlShape,
): FloatingToolbarItem<Command>[] => {
  const { floatingToolbarLinkSettingsButton } = getFeatureFlags(state);
  return floatingToolbarLinkSettingsButton === 'true'
    ? [
        {
          id: 'editor.link.settings',
          type: 'button',
          icon: CogIcon,
          title: intl.formatMessage(linkToolbarMessages.settingsLink),
          onClick: openLinkSettings,
        },
        { type: 'separator' },
      ]
    : [];
};
