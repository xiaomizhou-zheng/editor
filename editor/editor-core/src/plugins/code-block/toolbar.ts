import { defineMessages } from 'react-intl-next';
import RemoveIcon from '@atlaskit/icon/glyph/editor/remove';
import CopyIcon from '@atlaskit/icon/glyph/copy';
import { findDomRefAtPos } from 'prosemirror-utils';
import {
  FloatingToolbarHandler,
  FloatingToolbarButton,
  FloatingToolbarSeparator,
  FloatingToolbarListPicker,
} from '../floating-toolbar/types';
import {
  removeCodeBlock,
  changeLanguage,
  copyContentToClipboard,
  resetCopiedState,
} from './actions';
import commonMessages from '../../messages';
import { codeBlockCopyButtonMessages } from '@atlaskit/editor-common/messages';
import { CodeBlockState } from './pm-plugins/main-state';
import {
  provideVisualFeedbackForCopyButton,
  removeVisualFeedbackForCopyButton,
} from './pm-plugins/codeBlockCopySelectionPlugin';
import { Command } from '../../types';
import { hoverDecoration } from '../base/pm-plugins/decoration';
import { pluginKey } from './plugin-key';
import { SelectOption } from '../floating-toolbar/ui/Select';
import {
  createLanguageList,
  getLanguageIdentifier,
  DEFAULT_LANGUAGES,
} from './language-list';

export const messages = defineMessages({
  selectLanguage: {
    id: 'fabric.editor.selectLanguage',
    defaultMessage: 'Select language',
    description:
      'Code blocks display software code. A prompt to select the software language the code is written in.',
  },
});

const languageList = createLanguageList(DEFAULT_LANGUAGES);

export const getToolbarConfig = (
  allowCopyToClipboard: boolean = false,
): FloatingToolbarHandler => (state, { formatMessage }) => {
  const codeBlockState: CodeBlockState | undefined = pluginKey.getState(state);
  const pos = codeBlockState?.pos ?? null;

  if (!codeBlockState || pos === null) {
    return;
  }

  const node = state.doc.nodeAt(pos);
  const nodeType = state.schema.nodes.codeBlock;

  if (node?.type !== nodeType) {
    return;
  }

  const language = node?.attrs?.language;

  const options = languageList.map((lang) => ({
    label: lang.name,
    value: getLanguageIdentifier(lang),
    alias: lang.alias,
  }));

  // If language is not undefined search for it in the value and then search in the aliases
  const defaultValue = language
    ? options.find((option) => option.value === language) ||
      options.find((option) => option.alias.includes(language as never))
    : null;

  const languageSelect: FloatingToolbarListPicker<Command> = {
    id: 'editor.codeBlock.languageOptions',
    type: 'select',
    selectType: 'list',
    onChange: (option) => changeLanguage(option.value),
    defaultValue,
    placeholder: formatMessage(messages.selectLanguage),
    options,
    filterOption: languageListFilter,
  };

  const separator: FloatingToolbarSeparator = {
    type: 'separator',
  };

  const copyToClipboardItems = !allowCopyToClipboard
    ? []
    : ([
        {
          id: 'editor.codeBlock.copy',
          type: 'button',
          appearance: 'subtle',
          icon: CopyIcon,
          // note: copyContentToClipboard contains logic that also removes the
          // visual feedback for the copy button
          onClick: copyContentToClipboard,
          title: formatMessage(
            codeBlockState.contentCopied
              ? codeBlockCopyButtonMessages.copiedCodeToClipboard
              : codeBlockCopyButtonMessages.copyCodeToClipboard,
          ),
          onMouseEnter: provideVisualFeedbackForCopyButton,
          // note: resetCopiedState contains logic that also removes the
          // visual feedback for the copy button
          onMouseLeave: resetCopiedState,
          onFocus: provideVisualFeedbackForCopyButton,
          onBlur: removeVisualFeedbackForCopyButton,
          hideTooltipOnClick: false,
          disabled: codeBlockState.isNodeSelected,
          tabIndex: null,
        },
        separator,
      ] as const);

  const deleteButton: FloatingToolbarButton<Command> = {
    id: 'editor.codeBlock.delete',
    type: 'button',
    appearance: 'danger',
    icon: RemoveIcon,
    onMouseEnter: hoverDecoration(nodeType, true),
    onMouseLeave: hoverDecoration(nodeType, false),
    onFocus: hoverDecoration(nodeType, true),
    onBlur: hoverDecoration(nodeType, false),
    onClick: removeCodeBlock,
    title: formatMessage(commonMessages.remove),
    tabIndex: null,
  };

  return {
    title: 'CodeBlock floating controls',
    getDomRef: (view) =>
      findDomRefAtPos(pos, view.domAtPos.bind(view)) as HTMLElement,
    nodeType,
    items: [languageSelect, separator, ...copyToClipboardItems, deleteButton],
    scrollable: true,
  };
};

/**
 * Filters language list based on both name and alias properties.
 */
export const languageListFilter = (option: SelectOption, rawInput: string) => {
  const { data } = option as any;
  const searchString = rawInput.toLowerCase();
  return (
    data.label.toLowerCase().includes(searchString) ||
    data.alias.some((alias: string) => alias.toLowerCase() === searchString)
  );
};
