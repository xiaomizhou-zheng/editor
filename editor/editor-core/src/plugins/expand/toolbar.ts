import RemoveIcon from '@atlaskit/icon/glyph/editor/remove';

import commonMessages from '../../messages';
import { FloatingToolbarHandler } from '../floating-toolbar/types';
import { deleteExpand } from './commands';
import { hoverDecoration } from '../base/pm-plugins/decoration';
import { getPluginState } from './pm-plugins/plugin-factory';

export const getToolbarConfig: FloatingToolbarHandler = (
  state,
  { formatMessage },
) => {
  const { expandRef } = getPluginState(state);
  if (expandRef) {
    const { nestedExpand, expand } = state.schema.nodes;
    return {
      title: 'Expand toolbar',
      getDomRef: () => expandRef,
      nodeType: [nestedExpand, expand],
      offset: [0, 6],
      items: [
        {
          type: 'copy-button',
          items: [
            {
              state,
              formatMessage,
              nodeType: [nestedExpand, expand],
            },
            {
              type: 'separator',
            },
          ],
        },
        {
          id: 'editor.expand.delete',
          type: 'button',
          appearance: 'danger',
          icon: RemoveIcon,
          onClick: deleteExpand(),
          onMouseEnter: hoverDecoration([nestedExpand, expand], true),
          onMouseLeave: hoverDecoration([nestedExpand, expand], false),
          onFocus: hoverDecoration([nestedExpand, expand], true),
          onBlur: hoverDecoration([nestedExpand, expand], false),
          title: formatMessage(commonMessages.remove),
          tabIndex: null,
        },
      ],
    };
  }
  return;
};
