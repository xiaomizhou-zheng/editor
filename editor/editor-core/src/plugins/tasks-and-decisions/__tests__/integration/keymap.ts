import { BrowserTestCase } from '@atlaskit/webdriver-runner/runner';
import WebdriverPage from '@atlaskit/webdriver-runner/wd-wrapper';
import {
  goToEditorTestingWDExample,
  mountEditor,
} from '@atlaskit/editor-test-helpers/testing-example-page';
import {
  clickFirstCell,
  tableSelectors,
} from '@atlaskit/editor-test-helpers/page-objects/table';
import {
  expectToMatchSelection,
  editable,
  getDocFromElement,
  animationFrame,
} from '@atlaskit/editor-test-helpers/integration/helpers';
import { clickTaskNth } from '@atlaskit/editor-test-helpers/page-objects/task';
import { clickNthDecision } from '@atlaskit/editor-test-helpers/page-objects/decision';
import { fullpage } from '@atlaskit/editor-test-helpers/integration/helpers';
import taskListTableAdf from '../__fixtures__/tasklist-in-table-adf.json';
import decisionListInTableAdf from '../__fixtures__/decisionlist-in-table-adf.json';
import { retryUntilStablePosition } from '@atlaskit/editor-test-helpers/page-objects/toolbar';

const initEditor = async (page: WebdriverPage, adf: Object) =>
  await mountEditor(page, {
    appearance: fullpage.appearance,
    defaultValue: adf,
    allowTables: {},
    allowNestedTasks: true,
  });

// Chrome webdriver holds shift down, whilst firefox and edge just press it
const subsequentShiftTab = async (page: WebdriverPage) => {
  if (page.isBrowser('chrome')) {
    await page.keys(['Tab']);
  } else {
    await page.keys(['Shift', 'Tab'], true);
  }
};

// ALl tests skipped in safari as webdriver was encountering session termination issues
BrowserTestCase(
  'keymap.ts: tabbing from the first taskItem in a tableCell should go to the next cell',
  { skip: ['safari'] },
  async (client: any) => {
    const page = await goToEditorTestingWDExample(client);
    await initEditor(page, taskListTableAdf);

    await clickTaskNth(page, 1);
    await page.keys(['Tab']);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 62,
      to: 66,
    });
  },
);

BrowserTestCase(
  'keymap.ts: shift+tab from the first taskItem in a tableCell should go to the previous cell',
  { skip: ['safari'] },
  async (client: any) => {
    const page = await goToEditorTestingWDExample(client);
    await initEditor(page, taskListTableAdf);

    await clickTaskNth(page, 1);
    await page.keys(['Shift', 'Tab'], true);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 18,
      to: 22,
    });

    if (page.isBrowser('chrome')) {
      // unpress shift for the next test
      await page.keys(['Shift']);
    }
  },
);

BrowserTestCase(
  'keymap.ts: tab and shift+tab respectively indent and unident subsequent actionItems in a table',
  { skip: ['safari'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingWDExample(client);
    await initEditor(page, taskListTableAdf);

    // tab to indent
    await clickTaskNth(page, 2);
    await page.keys(['Tab']);
    await animationFrame(page);
    expect(
      await page.$eval(editable, getDocFromElement),
    ).toMatchCustomDocSnapshot(`${testName}:indent`);

    // shift + tab to unindent
    await page.keys(['Shift', 'Tab'], true);
    await animationFrame(page);
    expect(
      await page.$eval(editable, getDocFromElement),
    ).toMatchCustomDocSnapshot(`${testName}:unindent`);

    if (page.isBrowser('chrome')) {
      // unpress shift for the next test
      await page.keys(['Shift']);
    }
  },
);

BrowserTestCase(
  'keymap.ts: can tab back and forth through a table cell containing a taskList',
  { skip: ['safari'] },
  async (client: any) => {
    const page = await goToEditorTestingWDExample(client);
    await initEditor(page, taskListTableAdf);

    await clickFirstCell(page, true);
    // need to wait for cursor to be stable in the cell
    await retryUntilStablePosition(
      page,
      async () => {
        await page.waitForSelector(tableSelectors.contextualMenuButton);
      },
      tableSelectors.contextualMenuButton,
    );

    // tab forward
    await page.keys(['Tab']);
    await animationFrame(page);
    await page.keys(['Tab']);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 62,
      to: 66,
    });

    // tab backward
    await page.keys(['Shift', 'Tab'], true);
    await animationFrame(page);
    await subsequentShiftTab(page);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 18,
      to: 22,
    });

    if (page.isBrowser('chrome')) {
      // unpress shift for the next test
      await page.keys(['Shift']);
    }
  },
);

BrowserTestCase(
  'keymap.ts: tabbing from any decisionItem in a tableCell should go to the next cell',
  { skip: ['safari'] },
  async (client: any) => {
    const page = await goToEditorTestingWDExample(client);
    await initEditor(page, decisionListInTableAdf);

    await clickNthDecision(page, 1);
    await page.keys(['Tab']);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 64,
      to: 68,
    });

    await clickNthDecision(page, 2);
    await page.keys(['Tab']);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 64,
      to: 68,
    });
  },
);

BrowserTestCase(
  'keymap.ts: shift+tabbing from any decisionItem in a tableCell should go to the previous cell',
  { skip: ['safari'] },
  async (client: any) => {
    const page = await goToEditorTestingWDExample(client);
    await initEditor(page, decisionListInTableAdf);

    await clickNthDecision(page, 1);
    await page.keys(['Shift', 'Tab'], true);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 18,
      to: 22,
    });

    await clickNthDecision(page, 2);
    await subsequentShiftTab(page);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 18,
      to: 22,
    });

    if (page.isBrowser('chrome')) {
      // unpress shift for the next test
      await page.keys(['Shift']);
    }
  },
);

BrowserTestCase(
  'keymap.ts: can tab back and forth through a table cell containing a decisionList',
  { skip: ['safari'] },
  async (client: any) => {
    const page = await goToEditorTestingWDExample(client);
    await initEditor(page, decisionListInTableAdf);

    await clickFirstCell(page, true);
    // need to wait for cursor to be stable in the cell
    await retryUntilStablePosition(
      page,
      async () => {
        await page.waitForSelector(tableSelectors.contextualMenuButton);
      },
      tableSelectors.contextualMenuButton,
    );

    // tab forward
    await page.keys(['Tab']);
    await animationFrame(page);
    await page.keys(['Tab']);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 64,
      to: 68,
    });

    // tab backward
    await page.keys(['Shift', 'Tab'], true);
    await animationFrame(page);
    await subsequentShiftTab(page);
    await animationFrame(page);
    await expectToMatchSelection(page, {
      type: 'text',
      from: 18,
      to: 22,
    });
  },
);
