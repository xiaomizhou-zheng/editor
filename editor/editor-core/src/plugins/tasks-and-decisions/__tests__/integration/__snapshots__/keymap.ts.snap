// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keymap.ts: tab and shift+tab respectively indent and unident subsequent actionItems in a table:indent 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__autoSize": false,
        "isNumberColumnEnabled": false,
        "layout": "default",
        "localId": "abc-123",
      },
      "content": Array [
        Object {
          "content": Array [
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableHeader",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableHeader",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableHeader",
            },
          ],
          "type": "tableRow",
        },
        Object {
          "content": Array [
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "text": "text",
                      "type": "text",
                    },
                  ],
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "attrs": Object {
                    "localId": "abc-123",
                  },
                  "content": Array [
                    Object {
                      "attrs": Object {
                        "localId": "abc-123",
                        "state": "TODO",
                      },
                      "content": Array [
                        Object {
                          "text": "Action item 1",
                          "type": "text",
                        },
                      ],
                      "type": "taskItem",
                    },
                    Object {
                      "attrs": Object {
                        "localId": "abc-123",
                      },
                      "content": Array [
                        Object {
                          "attrs": Object {
                            "localId": "abc-123",
                            "state": "TODO",
                          },
                          "content": Array [
                            Object {
                              "text": "Action item 2",
                              "type": "text",
                            },
                          ],
                          "type": "taskItem",
                        },
                      ],
                      "type": "taskList",
                    },
                  ],
                  "type": "taskList",
                },
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "text": "text",
                      "type": "text",
                    },
                  ],
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
          ],
          "type": "tableRow",
        },
        Object {
          "content": Array [
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
          ],
          "type": "tableRow",
        },
      ],
      "type": "table",
    },
  ],
  "type": "doc",
}
`;

exports[`keymap.ts: tab and shift+tab respectively indent and unident subsequent actionItems in a table:unindent 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__autoSize": false,
        "isNumberColumnEnabled": false,
        "layout": "default",
        "localId": "abc-123",
      },
      "content": Array [
        Object {
          "content": Array [
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableHeader",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableHeader",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableHeader",
            },
          ],
          "type": "tableRow",
        },
        Object {
          "content": Array [
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "text": "text",
                      "type": "text",
                    },
                  ],
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "attrs": Object {
                    "localId": "abc-123",
                  },
                  "content": Array [
                    Object {
                      "attrs": Object {
                        "localId": "abc-123",
                        "state": "TODO",
                      },
                      "content": Array [
                        Object {
                          "text": "Action item 1",
                          "type": "text",
                        },
                      ],
                      "type": "taskItem",
                    },
                    Object {
                      "attrs": Object {
                        "localId": "abc-123",
                        "state": "TODO",
                      },
                      "content": Array [
                        Object {
                          "text": "Action item 2",
                          "type": "text",
                        },
                      ],
                      "type": "taskItem",
                    },
                  ],
                  "type": "taskList",
                },
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "text": "text",
                      "type": "text",
                    },
                  ],
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
          ],
          "type": "tableRow",
        },
        Object {
          "content": Array [
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "attrs": Object {
                "background": null,
                "colspan": 1,
                "colwidth": null,
                "rowspan": 1,
              },
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "tableCell",
            },
          ],
          "type": "tableRow",
        },
      ],
      "type": "table",
    },
  ],
  "type": "doc",
}
`;
